File: Platforms/Common/Reign.Video.XNA/States/BlendStateDesc.cs
using System;
using Reign.Core;
using X = Microsoft.Xna.Framework.Graphics;

namespace Reign.Video.XNA
{
	public class BlendStateDesc : BlendStateDescI
	{
		#region Properties
		internal X.ColorWriteChannels renderTargetWriteMask;

		internal bool blendEnable;
		internal X.BlendFunction blendOp;
		internal X.Blend srcBlend;
		internal X.Blend dstBlend;

		internal X.BlendFunction blendOpAlpha;
		internal X.Blend srcBlendAlpha;
		internal X.Blend dstBlendAlpha;
		#endregion

		#region Constructors
		public static BlendStateDesc New(BlendStateTypes type)
		{
			return new BlendStateDesc(type);
		}

		public BlendStateDesc(BlendStateTypes type)
		{
			renderTargetWriteMask = X.ColorWriteChannels.All;

			switch (type)
			{
			    case (BlendStateTypes.None):
			        blendEnable = false;
					blendOp = X.BlendFunction.Add;
					srcBlend = X.Blend.One;
					dstBlend = X.Blend.One;

					blendOpAlpha = X.BlendFunction.Add;
					srcBlendAlpha = X.Blend.One;
					dstBlendAlpha = X.Blend.One;
			        break;

				case (BlendStateTypes.Add):
			        blendEnable = true;
					blendOp = X.BlendFunction.Add;
					srcBlend = X.Blend.One;
					dstBlend = X.Blend.One;

					blendOpAlpha = X.BlendFunction.Add;
					srcBlendAlpha = X.Blend.One;
					dstBlendAlpha = X.Blend.One;
			        break;

				case (BlendStateTypes.Subtract):
			        blendEnable = true;
					blendOp = X.BlendFunction.Subtract;
					srcBlend = X.Blend.One;
					dstBlend = X.Blend.One;

					blendOpAlpha = X.BlendFunction.Subtract;
					srcBlendAlpha = X.Blend.One;
					dstBlendAlpha = X.Blend.One;
			        break;

				case (BlendStateTypes.RevSubtract):
			        blendEnable = true;
					blendOp = X.BlendFunction.ReverseSubtract;
					srcBlend = X.Blend.One;
					dstBlend = X.Blend.One;

					blendOpAlpha = X.BlendFunction.ReverseSubtract;
					srcBlendAlpha = X.Blend.One;
					dstBlendAlpha = X.Blend.One;
			        break;

				case (BlendStateTypes.Alpha):
			        blendEnable = true;
					blendOp = X.BlendFunction.Add;
					srcBlend = X.Blend.SourceAlpha;
					dstBlend = X.Blend.InverseSourceAlpha;

					blendOpAlpha = X.BlendFunction.Add;
					srcBlendAlpha = X.Blend.SourceAlpha;
					dstBlendAlpha = X.Blend.InverseSourceAlpha;
			        break;

				default:
					Debug.ThrowError("BlendStateDesc", "Unsuported BlendStateType");
					break;
			}
		}
		#endregion
	}
}

File: Platforms/Windows/Reign_Video_D3D9_Component/BlendStateDesc.cpp
#include "pch.h"
#include "BlendState.h"

namespace Reign_Video_D3D9_Component
{
	BlendStateDescCom::BlendStateDescCom(DWORD renderTargetWriteMask, bool blendEnable, REIGN_D3DBLENDOP blendOp, REIGN_D3DBLEND srcBlend, REIGN_D3DBLEND dstBlend, bool blendEnableAlpha, REIGN_D3DBLENDOP blendOpAlpha, REIGN_D3DBLEND srcBlendAlpha, REIGN_D3DBLEND dstBlendAlpha)
	{
		this->renderTargetWriteMask = renderTargetWriteMask;

		this->blendEnable = blendEnable;
		this->blendOp = (D3DBLENDOP)blendOp;
		this->srcBlend = (D3DBLEND)srcBlend;
		this->dstBlend = (D3DBLEND)dstBlend;

		this->blendEnableAlpha = blendEnableAlpha;
		this->blendOpAlpha = (D3DBLENDOP)blendOpAlpha;
		this->srcBlendAlpha = (D3DBLEND)srcBlendAlpha;
		this->dstBlendAlpha = (D3DBLEND)dstBlendAlpha;
	}
}